accepted/future-releases/static-extension-methods/design-document.md

```diff
@@ -0,0 +1,482 @@
+
+  - the extension name is visible in the lexical scope,
+  - the extension declares an instance member with the same base name, and 
+  - the `on` type (after type inference) of the extension is a super-type of the static type of the receiver.
+
+- Type inference for `extension Foo<T> on Bar<T> { baz<S>(params) => ...}` for an invocation `receiver.baz(args)` is performed as if the extension was a class:
+
+  ```dart
+  class Foo<T> {
+    Bar<T> _receiver;
+    Foo(Bar<T> this._receiver);
+    void baz<S>(params) => ...;
+  }
+  ```
+
+  that was invoked as `Foo(receiver).baz(args)`. The binding of `T` and `S` found here is the same binding used by the extension.  If the constructor invocation would be a compile-time error, the extension does not apply.
+
+- One extension is more specific than another if the instantiated `on` type of the former is a proper subtype of the instantiated `on` type of the latter, or if the two instantiated types are equivalent and the instantiate-to-bounds `on` type of the former is a proper subtype of the one on the latter.  An `on` type `T?`, with a trailing `?` works like a NNBD nullable type.
+
+- If there is no single most-specific extension which applies to a member invocation, then it is a compile-time error. (This includes the case with no applicable extensions, which is just the current behavior).
+
+- Otherwise, the single most-specific extension's member is invoked with the extension's type parameters bound to the types found by inference, and with `this ` bound to the receiver.
+
+- An extension method can be invoked explicitly using the syntax `ExtensionName(object).method(args)`. Type arguments can be applied to the extension explicitly as well, `MyList<String>(listOfString).quickSort()`. Such an invocation overrides all extension resolution. It is a compile-time error if `ExtensionName` would not apply to the `object.method(args)` invocation if it was in scope.  
+
+- The override can also be used for extensions imported with a prefix (which are not otherwise in scope): `prefix.ExtensionName(object).method(args)`.
+
+- An invocation of an extension method throws if the receiver is `null` unless the `on` type has a trailing `?`. With NNBD types, the invocation throws if the receiver is `null` and the instantiated `on` type of the selected extension does not accept `null`. (In most cases, this case can be excluded statically, but not for unsafely nullable types like `int*`).
+
+- Otherwise an invocation of an extension method runs the instance method with `this` bound to the receiver and with type variables bound to the types found by type inference (or written explicitly for an override invocation). The static type of `this` is the `on` type of the extension.
+
+- Inside an instance extension member, extension members accessed by unqualified name are treated as extension override accesses on `this`. Otherwise invocations on `this` are treated as any other invocations on the same static type.
```

Comment:
As mentioned in comment on line 263, I believe that this rule would be consistent with the approach proposed in #328, but not consistent with the current rules for the treatment of unqualified identifiers and `this`.

In short, the current rules stipulate that an unqualified name `m` that resolves to an instance member in the lexical scope will _not_ be taken to denote that instance member directly, it will be desugared to `this.m` and then processed as such.